Solutions/Corvus.Testing.AzureFunctions.Xunit.Demo/DemoFunctionFacts.cs

```diff
@@ -0,0 +1,63 @@
+// <copyright file="DemoFunctionFacts.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Corvus.Testing.AzureFunctions.Xunit.Demo
+{
+    using System.Net;
+    using System.Net.Http;
+    using System.Text;
+    using System.Threading.Tasks;
+    using global::Xunit;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Linq;
+
+    public abstract class DemoFunctionFacts
+    {
+        private readonly HttpClient client;
+        private HttpResponseMessage? lastHttpResponseMessage;
+
+        protected DemoFunctionFacts()
+        {
+            this.client = new HttpClient();
+        }
+
+        protected async Task When_I_GET(string uri)
+        {
+            this.lastHttpResponseMessage = await this.client.GetAsync(uri).ConfigureAwait(false);
+        }
+
+        protected async Task When_I_POST(string uri, object? content = null)
+        {
+            HttpContent? requestBody = null;
+            if (content != null)
+            {
+                requestBody = new StringContent(
+                    JObject.FromObject(content).ToString(Formatting.None),
+                    Encoding.UTF8,
+                    "application/json");
+            }
+
+            await this.IssuePostRequest(uri, requestBody);
+        }
+
+        protected void Then_I_receive(HttpStatusCode expected)
+        {
+            Assert.NotNull(this.lastHttpResponseMessage);
+            Assert.Equal(expected, this.lastHttpResponseMessage!.StatusCode);
```

Comment:
It's possible for methods like `Assert.NotNull` to include annotations that let the compiler know that an argument definitely won't be null if the method returns without error, so in a fully nullable-types-aware world, that `!` should not be necessary in the `Assert.Equal` line, because the compiler would (given a properly-annotated `Assert.NotNull` method) correctly infer that if execution makes it to this line, `this.lastHttpResponseMessage` cannot be null.

I'm assuming this `Assert` implementation comes from Xunit. Have they not updated their library for C# 8.0 nullable types?

Reply:
It would appear they have not yet, no. There's this [merged PR](https://github.com/xunit/assert.xunit/pull/36) on the assertion library which can be pulled in as a submodule. There's [no timescale at the moment](https://github.com/xunit/assert.xunit/pull/36#issuecomment-580461126) for a release with this change in. The [PR adding support to the core library](https://github.com/xunit/xunit/pull/2044) is going stale.